pyiron/atomistics/master/elastic.py

```diff
@@ -0,0 +1,385 @@
+        else:
+            E = E[:,na]**(np.arange(cc)+3)[na,:]
+        starting_index = np.sum(np.any(strain_higher_terms!=0, axis=0))
+        strain_higher_terms[ind, starting_index:starting_index+E.shape[1]] = E
+    strain_higher_terms = np.einsum('n,ij->nij',
+                                    np.ones(len(rotations)),
+                                    strain_higher_terms)
+    strain_higher_terms = strain_higher_terms.reshape(
+        -1, strain_higher_terms.shape[-1]
+    )
+    return strain_higher_terms
+
+def calc_elastic_tensor(
+        strain,
+        stress=None,
+        energy=None,
+        volume=None,
+        rotations=None,
+        return_score=False,
+        max_polynomial_order=None,
+        fit_first_order=False,
+    ):
+    """
+    Calculate 6x6-elastic tensor from the strain and stress or strain and
+    energy+volume.
+
+    Rotations matrices can be added to take box symmetries into account (unit
+    matrix can be added but does not have to be included in the list)
+
+    Args:
+        strain (numpy.ndarray): nx3x3 strain tensors
+        stress (numpy.ndarray): nx3x3 stress tensors
+        energy (numpy.ndarray): n energy values
+        volume (numpy.ndarray): n volume values
+        rotations (numpy.ndarray): mx3x3 rotation matrices
+        return_score (numpy.ndarray): return regression score
+            (cf. sklearn.linear_mode.LinearRegression)
+    """
+    if len(strain)==0:
+        raise ValueError('Not enough points')
+    if rotations is not None:
+        rotations = np.append(np.eye(3), rotations).reshape(-1, 3, 3)
+        _, indices = np.unique(
+            np.round(rotations, 6), axis=0, return_inverse=True
+        )
+        rotations = rotations[np.unique(indices)]
+    else:
+        rotations = np.eye(3).reshape(1, 3, 3)
+    if stress is not None and len(stress)==len(strain):
+        coeff, score = _fit_coeffs_with_stress(
+            strain=strain,
+            stress=stress,
+            rotations=rotations,
+            max_polynomial_order=max_polynomial_order,
+            fit_first_order=fit_first_order,
+        )
+    elif (energy is not None
+          and volume is not None
+          and len(energy)==len(strain)
+          and len(energy)==len(volume)):
+        coeff, score = _fit_coeffs_with_energies(
+            strain=strain,
+            energy=energy,
+            volume=volume,
+            rotations=rotations,
+            max_polynomial_order=max_polynomial_order,
+            fit_first_order=fit_first_order,
+        )
+    else:
+        raise ValueError('Provide either strain and stress, or strain, energy '
+                         + 'and volume of same length.')
+    if return_score:
+        return np.array(coeff)[:,:6], score
+    else:
+        return np.array(coeff)[:,:6]
+
+def get_strain(max_strain=0.05, n_set=10, polynomial_order=3, additional_points=0, normalize=False):
+    strain_lst = np.random.random((n_set, 3, 3))-0.5
```

Comment:
This code also appears down in `ElasticTensor._create_strain_matrices`. I think it would be worthwhile factoring out a method

```python
_random_symmetric_matrices(n):
    matrices = np.random.random((n, 3, 3))-0.5
    return matrices + np.einsum('nij->nji', matrices)
```

Here and below you rescale, but here you also (potentially) normalize. I think it's fine to just return random on (-0.5, 0.5) and rescale similar to you have it. Although later you can do something like `eps = self.scale * _random_symmetric_matrices(self.n_whatevers)`. (But with all the right names, obviously.)

Reply:
p.s. I think I might be learning einsum by osmosis 😝